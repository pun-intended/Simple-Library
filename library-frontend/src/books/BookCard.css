img {
    max-width: 95%;
    max-height: 95%;
    object-fit: contain;
}


.bookRow {
    align-items: stretch;
}

.bookTitle {
    padding: 15px 3px 5px;
    padding-right: 3px;
}

.bookTitle h4 {
    font-size: .7em;
}

.bookImage {
    justify-content: center;
    align-items: center;
}

.bookListBtn {
     width: 6.2em;
     padding: 4px 0px 4px 0px;
     font-size: .7em;
}