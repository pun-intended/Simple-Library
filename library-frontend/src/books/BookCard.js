// Simple card for list
// Cover (l, set height), title, who has, borrowed date
// OnClick function pass
import React, { useState } from "react"
import {Card, CardBody, CardTitle, CardText, Button } from 'reactstrap'
import CheckOutModal from "./CheckOutModal.js";
import CheckInModal from "./CheckInModal.js";

const BookCard = ({book, setUpdate}) => {

    const [outModal, setOutModal] = useState(false);
    const [inModal, setInModal] = useState(false);
    const toggleOut = () => { setOutModal(!outModal)};
    const toggleIn = () => { setInModal(!inModal)};

    return(
        <Card >
            <CardBody>
            <CardTitle tag='h5'>
                {`${book.title}`}
            </CardTitle>
            </CardBody >
            <CardBody >
                <img className="BookCard BookCover" 
                    src={`https://covers.openlibrary.org/b/isbn/${book.isbn}-M.jpg`} 
                    alt={`${book.title}`} />
            </CardBody>
            <CardBody >
                <CardText>
                    {!book.available && <Button color="secondary" className="btn-primary" onClick={toggleIn}>Check-in</Button>}
                    {book.available && <Button color="primary" className="btn-primary" onClick={toggleOut}>Check-out</Button>}
                </CardText>
            </CardBody>
           {outModal && <CheckOutModal modal={outModal} toggle={toggleOut} book={book} setUpdate={setUpdate}/>}
           {inModal && <CheckInModal modal={inModal} toggle={toggleIn} book_id={book.id} setUpdate={setUpdate}/>}
        </Card>
    )
}

export default BookCard