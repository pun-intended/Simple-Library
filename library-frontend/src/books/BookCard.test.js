import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import BookCard from "./BookCard.css";
import StudentContext from "../StudentContext";


const bookIn = {
    id: "101",
    title: "test book - in",
    isbn: "0-330-25864-8",
    stage: 2,
    condition: "great",
    available: true,
};

const bookOut = {
    id: "102",
    title: "test book - out",
    isbn: "	0-307-26543-9",
    stage: 2,
    condition: "good",
    available: false,
};

const students = {
    students: [{
        fist_name: "test",
        last_name: "student",
    },
    {
        fist_name: "test",
        last_name: "student",
    }]
};

const setUpdate = jest.fn();

// Smoke test
it("renders without crashing", () => {
    render(<BookCard book={bookIn} setUpdate={setUpdate} />)
});

// Snapshot
const {asFragment} = render(<BookCard book={bookIn} setUpdate={setUpdate} />);
expect(asFragment()).toMatchSnapshot();

// Check-out button shown for available book
it("renders check-out button for available book", () => {
    const { getByText } = render(<BookCard book={bookIn} setUpdate={setUpdate} />);
    expect(getByText("Check-out")).toBeInTheDocument();
});

// Check-in button shown for unavailable book
it("renders check-in button for unavailable book", () => {
    const { getByText } = render(<BookCard book={bookOut} setUpdate={setUpdate} />);
    expect(getByText("Check-in")).toBeInTheDocument();
});

// check-in button opens check-in modal
it("opens check-in modal", () => {
    const { getAllByText } = render(<BookCard book={bookOut} setUpdate={setUpdate} />);
    const btn = getAllByText(/check-in/i);
    fireEvent.click(btn[0]); 
    expect(screen.getByRole('dialog')).toBeInTheDocument();
}); 

// check-out button opens check-out modal
it("opens check-out modal", () => {
    const { getAllByText } = render(
        <StudentContext.Provider value={students}>
        <BookCard book={bookIn} setUpdate={setUpdate} />
        </StudentContext.Provider>);
    const btn = getAllByText(/check-out/i);
    fireEvent.click(btn[0]); 
    expect(screen.getByRole('dialog')).toBeInTheDocument();
});