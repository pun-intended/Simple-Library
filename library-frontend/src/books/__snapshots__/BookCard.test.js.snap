// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="card-body"
    >
      <h5
        class="card-title"
      >
        test book - in
      </h5>
    </div>
    <div
      class="card-body"
    >
      <img
        alt="test book - in"
        class="BookCard BookCover"
        src="https://covers.openlibrary.org/b/isbn/0-330-25864-8-M.jpg"
      />
    </div>
    <div
      class="card-body"
    >
      <p
        class="card-text"
      >
        <button
          class="btn-primary btn btn-secondary"
          type="button"
        >
          Check-out
        </button>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
